### PR TITLE
[WIP] Delayed privatization.

### PR DESCRIPTION
**Pull request by ergawy**
This is a PoC for delayed privatization in OpenMP. Instead of directly
emitting privatization code in the frontend, we add a new op to outline
the privatization logic for a symbol and call-like mapping that maps
from the host symbol to an outlined function-like privatizer op.

Later, we would inline the delayed privatizer function-like op in the
OpenMP region to basically get the same code generated directly by the
fronend at the moment.

### This is my plan to productize this WIP (✅ means the item is already done, ⌛ means it is in progress):
1. In this PR: add a feature flag to control the usage of delayed privatization. This will be disabled by default and enabled gradually as we add support for delayed privatization for more cases. ✅ 
2. In this PR: handle the remaining comments so that you do not have to spend effort on them in later separate PRs.
3. This PR will **not** be merged since it is getting quite large. Instead I will split it into a number of PRs as explained in following points.
4. PR 1: a NFC PR to copy the refactoring of `genOpWithBody` introduced here. This will introduce the `genRegionEntryCB` callback and use it for worksharing loop ops. (#80817 #80839) ✅ 
5. PR 2: copy the `omp.private` op along with a roundtripping lit test to test parsing, printing, and verification. (#80955) ✅ 
6. PR 3: copy the extensions in `omp.parallel` op (i.e. adding the `private` clause) along with a roundtripping lit test as well. (#81452) ✅ 
7. PR 4: copy the changes introduced in the `OpenMPToLLVM.cpp` conversion pattern: `RegionOpConversion`. (#81414) ✅ 
9. PR 5: copy the changes introduced in `OpenMPToLLVMIRTranslation.cpp`: the `PrivCB` and `prepareOmpParallel`. (#81715) ✅ 
10. PR 6: copy the changes needed for creating the `omp.private` ops and modifying the creation of the `omp.parallel` op to use the new clause. These are most of the changes in `DataSharingProcessor` and `genParallelOp`. (https://github.com/llvm/llvm-project/pull/81833) ✅ 
11. PRs 7..N: Add delayed privatization support for more symbol types and more OMP ops (e.g. `target`). (https://github.com/llvm/llvm-project/pull/83761 https://github.com/llvm/llvm-project/pull/84123 https://github.com/llvm/llvm-project/pull/84296 https://github.com/llvm/llvm-project/pull/84740 https://github.com/llvm/llvm-project/pull/85023) ⌛ 
12. PR N+1: Remove the delayed privatization feature flag.

Please let me know if you disagree with anything.

### CODE REVIEW

File: flang/include/flang/Lower/AbstractConverter.h
Comment (ergawy) [inline flang/include/flang/Lower/AbstractConverter.h:1]: I had to augment quite a few methods on the `AbstractConvereter` API to add an external symbol map. This map in the case of outlined privatizers corresponds to the symbol map managed by the scope of the privatization op.

File: flang/lib/Lower/OpenMP.cpp [hunk 1]
```diff
@@ -487,15 +495,54 @@ void DataSharingProcessor::collectDefaultSymbols() {
   }
 }
 
-void DataSharingProcessor::privatize() {
+void DataSharingProcessor::privatize(
+    llvm::SetVector<mlir::omp::PrivateClauseOp> *privateInitializers) {
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP.cpp:499]: For now, we do nothing with `privateInitializaers`. However, the plan is to populate this list and use it to link the created ops with the corresponding block arguments for the private variables.

File: flang/include/flang/Lower/AbstractConverter.h
Comment (ergawy) [reply flang/include/flang/Lower/AbstractConverter.h:1]: Another option would be to treat `SymMap`s like `InsertionPoint`s in the sense that we always have an active `SymMap`/`InsertionPoint` at any given moment and we can save this `SymMap`/`InsertionPoint`, set a new one, and restore it later.

This way we have an API that replaces and restores the currently active `SymMap` within the converter and clean-up the API a little bit.

File: flang/lib/Lower/OpenMP.cpp (hunk 1 above)
Comment (ergawy) [reply flang/lib/Lower/OpenMP.cpp:499]: Thanks for taking a look!

Just to be sure we are on the same page, are you referring to using the push/pop scope mechanism instead of the `privatizerSymbolMap` I create in line 529 below?

The `privateInitializers` argument is not related to scoping and symbol map management. Its purpose is to be populated with the created ops so that we can use them as operands to the `omp.parallel`/`omp.target` ops (or any op with a `private` clause.

Comment (ergawy) [reply flang/lib/Lower/OpenMP.cpp:495]: The 2nd commit in the PR uses the push/pop scope mechanism. Maybe adding an getter to the symbol map is not the best approach (instead we might want to pass down the `Fortran::lower::SymMap &symTable` parameter from `genOMP`.

File: flang/include/flang/Lower/AbstractConverter.h [hunk 2]
```diff
@@ -295,6 +296,11 @@ class AbstractConverter {
     return loweringOptions;
   }
 
+  virtual Fortran::lower::SymbolBox
+  lookupOneLevelUpSymbol(const Fortran::semantics::Symbol &sym) = 0;
+
+  virtual Fortran::lower::SymMap *getLocalSymbols() { return nullptr; }
```
Comment (clementval) [inline flang/include/flang/Lower/AbstractConverter.h:302]: Can't you just pass the localSymbols as arguments to the OpenMP.cpp functions?

File: mlir/lib/Dialect/Func/IR/FuncOps.cpp [hunk 3]
```diff
@@ -138,9 +138,7 @@ LogicalResult ConstantOp::verify() {
   return success();
 }
 
-OpFoldResult ConstantOp::fold(FoldAdaptor adaptor) {
-  return getValueAttr();
-}
+OpFoldResult ConstantOp::fold(FoldAdaptor adaptor) { return getValueAttr(); }
```
Comment (clementval) [inline mlir/lib/Dialect/Func/IR/FuncOps.cpp:141]: Unrelated change?

File: mlir/test/Dialect/OpenMP/roundtrip.mlir [hunk 4]
```diff
@@ -0,0 +1,38 @@
+// RUN: fir-opt -verify-diagnostics %s | fir-opt | FileCheck %s
+
+// CHECK-LABEL: _QPprivate_clause
+func.func @_QPprivate_clause() {
+  %0 = fir.alloca i32 {bindc_name = "x", uniq_name = "_QFprivate_clause_allocatableEx"}
+  %1 = fir.alloca i32 {bindc_name = "y", uniq_name = "_QFprivate_clause_allocatableEy"}
+
+  // CHECK: omp.parallel private(@x.privatizer %0 -> %arg0, @y.privatizer %1 -> %arg1 : !fir.ref<i32>, !fir.ref<i32>)
+  omp.parallel private(@x.privatizer %0 -> %arg0, @y.privatizer %1 -> %arg1 : !fir.ref<i32>, !fir.ref<i32>) {
```
Comment (clementval) [inline mlir/test/Dialect/OpenMP/roundtrip.mlir:9]: Do you really need to add a block argument? The operands in  `private_vars` should already be block argument AFAICT.

File: flang/include/flang/Lower/SymbolMap.h [hunk 5]
```diff
@@ -101,6 +101,7 @@ struct SymbolBox : public fir::details::matcher<SymbolBox> {
                  [](const fir::FortranVariableOpInterface &x) {
                    return fir::FortranVariableOpInterface(x).getBase();
                  },
+                 [](const fir::MutableBoxValue &x) { return x.getAddr(); },
```
[the remark this answers is not in the document]
Comment (ergawy) [reply flang/include/flang/Lower/SymbolMap.h]: This was a debugging leftover :). Cleaned up things a bit more but there is still work to do in the PR of course.

File: mlir/test/Dialect/OpenMP/roundtrip.mlir (hunk 4 above)
Comment (ergawy) [reply mlir/test/Dialect/OpenMP/roundtrip.mlir]: >  The operands in `private_vars` should already be block argument AFAICT.

Not really. Value operands to an op do not translate automatically to arguments to its entry region. Please let me know if I misunderstood your comment :).

> Do you really need to add a block argument?

However, I agree with you that we probably do not need a block argument. We can simply keep references to the host region inside the OMP region and when we inline the privatizer later, we can inline and replace all references to the host value with the yielded value from the privatizer.

So the op would look like this before inlining:
```
    %0 = fir.alloca i32 {bindc_name = "var1", uniq_name = "_QFdelayed_privatizationEvar1"}
    %1 = fir.alloca i32 {bindc_name = "var2", uniq_name = "_QFdelayed_privatizationEvar2"}
    omp.parallel private(@var1.privatizer %0, @var2.privatizer %1 : !fir.ref<i32>, !fir.ref<i32>) {
      %2 = fir.load %0 : !fir.ref<i32>
      %3 = fir.load %1 : !fir.ref<i32>
      %4 = arith.addi %2, %3 : i32
      %c2_i32 = arith.constant 2 : i32
      %5 = arith.addi %4, %c2_i32 : i32
      fir.store %5 to %0 : !fir.ref<i32>
      omp.terminator
    }
```

@kiranchandramohan what do you think? Can we go with the above?

File: flang/lib/Lower/OpenMP.cpp [hunk 6]
```diff
@@ -487,15 +491,56 @@ void DataSharingProcessor::collectDefaultSymbols() {
...
+            converter.getLocalSymbols()->shallowLookupSymbol(*sym).getAddr());
+
+        converter.getLocalSymbols()->popScope();
+        converter.getLocalSymbols()->popScope();
+        firOpBuilder.restoreInsertionPoint(ip);
+      }
+
+      // TODO: This will eventually be an else to the `if` above it. For now, I
+      // emit both the outlined privatizer AND directly emitted cloning and
+      // copying ops while I am testing.
```
[the remark this answers is not in the document]
Comment (ergawy) [reply flang/lib/Lower/OpenMP.cpp]: I modified the `DataSharingProcessor` a bit to accommodate delayed privatization without adding a derived class. My argument is that it is simpler this way and a switch to control the privatization behavior is more appropriate.

Please let me know if you disagree :).

**Comment by ergawy** [comment]
@clementval @kiranchandramohan thanks a lot for taking a look. I think this WIP is getting closer and closer to having a complete end-to-end implementation. If I not missing anything, what remains is to actually inline the the privatizer in the op before converting to LLVM IR which I am looking into now.

Of course, I still need to support more complex symbols (e.g. allocatables) but I am focusing on simple variables for now to get something working end-to-end.

**Comment by kiranchandramohan** [comment]
> @clementval @kiranchandramohan thanks a lot for taking a look. I think this WIP is getting closer and closer to having a complete end-to-end implementation. If I not missing anything, what remains is to actually inline the the privatizer in the op before converting to LLVM IR which I am looking into now.
> 
> Of course, I still need to support more complex symbols (e.g. allocatables) but I am focusing on simple variables for now to get something working end-to-end.

Thanks for the work here. Forgot to mention that the OpenMPIRBuilder has a privatization call back that is designed to implement privatization. You can consider using that.
(https://github.com/llvm/llvm-project/blob/648eb7c1415afb818b45782e1f2758a1f7677496/llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp#L1544)

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td [hunk 7]
```diff
@@ -187,8 +188,10 @@ def ParallelOp : OpenMP_Op<"parallel", [
              Variadic<AnyType>:$allocate_vars,
              Variadic<AnyType>:$allocators_vars,
              Variadic<OpenMP_PointerLikeType>:$reduction_vars,
+             Variadic<AnyType>:$private_vars,
              OptionalAttr<SymbolRefArrayAttr>:$reductions,
-             OptionalAttr<ProcBindKindAttr>:$proc_bind_val);
+             OptionalAttr<ProcBindKindAttr>:$proc_bind_val,
+             OptionalAttr<SymbolRefArrayAttr>:$private_inits);
```
Comment (luporl) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:194]: Does each element of `private_inits` correspond to a PrivateClauseOp?
If this is the case, I think that a name like `private_ops` or `privatizers` would make it clearer what this attributes array is supposed to contain. The `_inits` suffix makes me think it holds the symbols from which values the privatized vars will be initialized.

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td [hunk 8]
```diff
@@ -187,8 +188,10 @@ def ParallelOp : OpenMP_Op<"parallel", [
              Variadic<AnyType>:$allocate_vars,
              Variadic<AnyType>:$allocators_vars,
              Variadic<OpenMP_PointerLikeType>:$reduction_vars,
+             Variadic<AnyType>:$private_vars,
```
Comment (luporl) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:191]: Is it possible to move `private_vars` to be right before `private_inits`?

File: flang/test/Lower/OpenMP/FIR/delayed_privatization.f90 [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+end subroutine
+
+! This is what flang emits with the PoC:
+! --------------------------------------
+!
+!func.func @_QPdelayed_privatization() {
+!  %0 = fir.alloca i32 {bindc_name = "var1", uniq_name = "_QFdelayed_privatizationEvar1"}
+!  %1 = fir.alloca i32 {bindc_name = "var2", uniq_name = "_QFdelayed_privatizationEvar2"}
+!  omp.parallel private(@var1.privatizer %0, @var2.privatizer %1 : !fir.ref<i32>, !fir.ref<i32>) {
```
Comment (luporl) [inline flang/test/Lower/OpenMP/FIR/delayed_privatization.f90:17]: I was expecting this to be
`omp.parallel private(@var1.privatizer %0 : !fir.ref<i32>, @var2.privatizer %1 : !fir.ref<i32>)`,
similar to what is done in `reduction` and `copyprivate`.
Is there any reason to list the types separately, at the end?

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td [hunk 10]
```diff
@@ -1478,6 +1485,38 @@ def Target_UpdateDataOp: OpenMP_Op<"target_update_data",
...
+  let summary = "TODO";
+  let description = [{}];
+
+  let arguments = (ins SymbolNameAttr:$sym_name,
+                       TypeAttrOf<FunctionType>:$function_type);
+
+  let regions = (region AnyRegion:$body);
+
+  let builders = [OpBuilder<(ins
+    "::mlir::Type":$privateVar,
```
Comment (luporl) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:1500]: Shouldn't it be `$privateVarType` here?

File: flang/test/Lower/OpenMP/FIR/delayed_privatization.f90 [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+!    %4 = arith.addi %2, %3 : i32
+!    %c2_i32 = arith.constant 2 : i32
+!    %5 = arith.addi %4, %c2_i32 : i32
+!    fir.store %5 to %0 : !fir.ref<i32>
+!    omp.terminator
+!  }
+!  return
+!}
+!
+!"omp.private"() <{function_type = (!fir.ref<i32>) -> !fir.ref<i32>, sym_name = "var1.privatizer"}> ({
```
Comment (luporl) [inline flang/test/Lower/OpenMP/FIR/delayed_privatization.f90:29]: Is it ok for the same `sym_name` to be repeated?
It's possible to have more than one `omp.private` operation with the same `sym_name`, if there are variables with the same name, in different scopes.

**Review by luporl — COMMENTED** [review]
Overall, your approach to implement delayed privatization looks good to me,
requiring only a minor modification to AbstractConverter/Bridge.

I just have a couple comments.

**Comment by ergawy** [comment]
> Forgot to mention that the OpenMPIRBuilder has a privatization call back that is designed to implement privatization. You can consider using that. (
> 
> https://github.com/llvm/llvm-project/blob/648eb7c1415afb818b45782e1f2758a1f7677496/llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp#L1544
> )

Thanks for pointing that out. Done, hopefully :).

File: flang/lib/Lower/OpenMP.cpp [hunk 12]
```diff
@@ -496,8 +513,46 @@ void DataSharingProcessor::privatize() {
         copyFirstPrivateSymbol(&*mem);
       }
     } else {
-      cloneSymbol(sym);
-      copyFirstPrivateSymbol(sym);
+      if (useDelayedPrivatizationWhenPossible) {
```
[the remark this answers is not in the document]
Comment (ergawy) [reply flang/lib/Lower/OpenMP.cpp:516]: > Is support for delayed privatization planned for common blocks too (in the if branch above)?

The plan is to indeed support all possible scenarios. I just started by prototyping the whole idea end-to-end (from FIR+OMP MLIR down to LLVM IR) which I just finished. So I will start fleshing out things a bit more now and look into supporting more variable types.

File: mlir/test/Dialect/OpenMP/roundtrip.mlir (hunk 4 above)
Comment (ergawy) [reply mlir/test/Dialect/OpenMP/roundtrip.mlir]: I ended up **not** using block arguments and directly mapping the host value to its private clone when lowering to LLVM (see the changes in `convertOmpParallel(...)`.

File: flang/lib/Lower/OpenMP.cpp [hunk 13]
```diff
@@ -161,6 +161,11 @@ class DataSharingProcessor {
   const Fortran::parser::OmpClauseList &opClauseList;
   Fortran::lower::pft::Evaluation &eval;
 
+  bool useDelayedPrivatizationWhenPossible;
```
[the remark this answers is not in the document]
Comment (ergawy) [reply flang/lib/Lower/OpenMP.cpp:164]: My initial idea was to **gradually** rollout support for delayed privatization so we would have a mix of eager and delayed privatization based on:
1. The type of the symbol being privatized
2. The operation supporting the `private` clause.
(that exclusion of unsupported symbol types/ops is not implemented yet, hence the test failures in the PR)

That's why I added then `WhenPossible` suffix (maybe this is not the best choice of words here). My aim is that eventually, we can get rid of eager privatization altogether and then there would be any need for this flag.

This is just a way to reduce the scope of the PR and not introduce a humongous change at once.

I hope that makes sense! Please let makes sense. Let me know if you disagree or have better suggestions.

File: mlir/test/Dialect/OpenMP/roundtrip.mlir (hunk 4 above)
Comment (kiranchandramohan) [reply mlir/test/Dialect/OpenMP/roundtrip.mlir]: Sorry, I missed this point.
I think we need a block argument or a new value to keep the dataflow correct in the MLIR layer. In the current representation, loads and stores to `%0` and `%1` are to the original variable and not the privatized variable.

Comment (ergawy) [reply mlir/test/Dialect/OpenMP/roundtrip.mlir]: No problem. Ok, that indeed makes sense. Will do ...

File: flang/lib/Lower/OpenMP.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (luporl) [reply flang/lib/Lower/OpenMP.cpp:164]: Ok, thanks for the clarification. If the exclusion of unsupported symbol types/ops is going to be added and this is meant to be a temporary transition flag, it should be fine.

File: mlir/test/Dialect/OpenMP/roundtrip.mlir (hunk 4 above)
Comment (ergawy) [reply mlir/test/Dialect/OpenMP/roundtrip.mlir]: Done. Unfortunately(?), I had to refactor `createBodyOfOp` a bit to do that. I hope the approach I took is sane enough :).

To see the current status of the whole pipeline, there is a file with an example `delayed_privatization.f90` in the PR. I will convert that file to a number of lit tests once the approach is settled.

**Comment by ergawy** [comment]
Thanks again for taking a look here. Since I see no objections to the taken approach (I hope I am not delusional :D), here is my plan:
1. In this PR: try to get the pipeline to be green and make sure that no lit tests fail because of the introduced changes.
2. In this PR: handle the remaining in of you comments so that you do not have to spend effort on them in later separate PRs.
3. This PR will **not** be merged since it is getting quite large. Instead I will split it into a number of PRs as explained in following points.
4. PR 1: a NFC PR to copy the refactoring of `genOpWithBody` introduced here. This will introduce the `genRegionEntryCB` callback and use it for worksharing loop ops.
5. PR 2: copy the `omp.private` op along with a roundtripping lit test to test parsing, printing, and verification.
6. PR 3: copy the extensions in `omp.parallel` op (i.e. adding the `private` clause) along with a roundtripping lit test as well.
7. PR 4: copy the changes introduced in the `OpenMPToLLVM.cpp` conversion pattern: `RegionOpConversion`.
8. PR 5: copy the changes introduced in `OpenMPToLLVMIRTranslation.cpp`: the `PrivCB` and `prepareOmpParallel`.
9. PR 6: copy the changes needed for creating the `omp.private` ops and modifying the creation of the `omp.parallel` op to use the new clause. These are most of the changes in `DataSharingProcessor` and `genParallelOp`.

Please let me know if you disagree with anything :).

File: flang/test/Lower/OpenMP/FIR/delayed_privatization.f90 (hunk 11 above)
Comment (ergawy) [reply flang/test/Lower/OpenMP/FIR/delayed_privatization.f90]: You are definitely correct. Repeating the symbol name produces invalid IR. I am working on it, I was just getting the basics working first :).

Comment (clementval) [reply flang/test/Lower/OpenMP/FIR/delayed_privatization.f90]: You can use the type to unique the privatizer. Use the getTypeAsString from FIRType.

**Comment by ergawy** [comment]
@kiranchandramohan @clementval @luporl can you 🙏 have a look at the proposed plan above and let me know if you agree with it? Apologies for the noise but I just want to make sure we are aligned since this touches a lot of places in the pipeline already.

I am specially interested in your opinion about the "feature flag" thing to enable/disable delayed privatization both in `bbc`and `flang-new` at well (and having it disabled by default). Let me know if this is something you agree with or have other suggestions so that we can roll out support for delayed privatization in a practical and gradual manner.

File: flang/lib/Lower/OpenMP.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (ergawy) [reply flang/lib/Lower/OpenMP.cpp]: I generalized the delayed privatization logic in a new method: `doPrivatize` and use the same logic for all paths that need privatization.

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td (hunk 8 above)
Comment (ergawy) [reply mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td]: Done.

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td (hunk 10 above)
Comment (ergawy) [reply mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td]: Done.

File: flang/test/Lower/OpenMP/FIR/delayed_privatization.f90 (hunk 11 above)
Comment (ergawy) [reply flang/test/Lower/OpenMP/FIR/delayed_privatization.f90]: > You can use the type to unique the privatizer. Use the getTypeAsString from FIRType.

Would that work if there are multiple parallel ops in the function that need 2 different privatizers; for example one `private(x)` and another `firstprivate(x)`? Or similarly if `x` is defined in more than one function?

I ended up using the `mlir::SymbolTable::generateSymbolName` utility. Let me know if you do not like this approach.

**Comment by kiranchandramohan** [comment]
> @kiranchandramohan @clementval @luporl can you 🙏 have a look at the proposed plan above and let me know if you agree with it? Apologies for the noise but I just want to make sure we are aligned since this touches a lot of places in the pipeline already.
> 
> I am specially interested in your opinion about the "feature flag" thing to enable/disable delayed privatization both in `bbc`and `flang-new` at well (and having it disabled by default). Let me know if this is something you agree with or have other suggestions so that we can roll out support for delayed privatization in a practical and gradual manner.

Thanks @ergawy for pushing this forward. This will add more capabilities to the dialect and make lowering easier. Your approach of gradually rolling out the feature makes a lot of sense. We should also ensure that applications that are enabled now continues to work (spec-2017 speed, subset of specomp2012, subset of NPB, SNAP) and that there is no regression in the gfortran testsuite.

When introducing the changes to OpenMP dialect, could you add a reply to https://discourse.llvm.org/t/rfc-privatisation-in-openmp-dialect/3526 saying that we are adding support for privatisation clauses in the dialect. We see benefits for this in:
-> lowering from frontend to MLIR
-> better support for task and target constructs. Tasks have delayed execution, so the private/firstprivate variables have to be inserted in the task structure. <Add some info for target as well>
-> better lowering (to LLVM) for index variables of loops.  https://github.com/llvm/llvm-project/pull/75836
-> better support for lastprivate, where the runtime can support detection of the last iteration.

I think it is fine to add a single flag to the file that can enable `delayed privatization` if it is accessible as `flang-new -mmlir <flag-name>` like for e.g (-fdebug-dump-pre-fir) and directly with bbc.

File: flang/test/Lower/OpenMP/FIR/delayed_privatization.f90 (hunk 11 above)
Comment (luporl) [reply flang/test/Lower/OpenMP/FIR/delayed_privatization.f90]: I think the idea would be to look for the privatizer in the module, using the name that can be obtained from the type of the variable. Then a new privatizer would be created only when none exists for the given type.
Can this be done with `omp.private`?

It is done for the copy functions, used with copyprivate:
https://github.com/llvm/llvm-project/pull/80485/files#diff-63581fb5be230c4918a0d8196a02ff696759f3010a419339c56a21777fb8c634R1221

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td (hunk 7 above)
Comment (ergawy) [reply mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td]: Done

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp [hunk 14]
```diff
@@ -3505,6 +3505,18 @@ struct ZeroOpConversion : public FIROpConversion<fir::ZeroOp> {
   }
 };
 
+class DeclareOpConversion : public FIROpConversion<fir::DeclareOp> {
```
Comment (ergawy) [inline flang/lib/Optimizer/CodeGen/CodeGen.cpp:3508]: While experimenting with delayed privation for `hlfir` (so far I have been experimenting with `fir`), I came across a small issue which is that `fir.declare` is not support when converting from `fir` to `mlir` dialect. That's why I had to copy this conversion pattern from `PreCGRewrite.cpp`. Since `DeclareOpConversion` does not convert to op in the `fir::cg` dialect, I think it would be nice to move it here to make conversion from `fir` to `llvm` dialect more complete.

Do you folks have any input on this?

Comment (ergawy) [reply flang/lib/Optimizer/CodeGen/CodeGen.cpp:3508]: Thanks @tblah for taking a look. I was referring to lowering using `fir-opt --fir-to-llvm-ir` for testing purposes. I expected that the `fir-opt --fir-to-llvm-ir` should provide a more-or-less complete conversion pipeline from `fir` to `llvm`. This is mostly to make testing and experimenting easier.

I am a bit of newbie to the project so I might not be using the different tools and their flags properly yet.

Comment (tblah) [reply flang/lib/Optimizer/CodeGen/CodeGen.cpp:3508]: No `fir-oppt --fir-to-llvm-ir` is not sufficient. The FIR pass pipeline is quite long (see https://github.com/llvm/llvm-project/blob/main/flang/test/Driver/mlir-pass-pipeline.f90). A few of these are optimizations (e.g. SimplifyIntrinsics, CSE, AliasTags), but several different passes are needed to convert FIR into LLVM (even more for HLFIR).

Roughly, to get from FIR to LLVM you need the passes in `createDefaultFIRCodeGenPassPipeline`: https://github.com/llvm/llvm-project/blob/ac3bd2bd530127786741bd9f164d66a3c3f40961/flang/include/flang/Tools/CLOptions.inc#L302

From the command line you can pipe FIR into `tco` and it will output LLVM IR. `fir-opt` is more for running individual passes. The help text is a bit hard to follow, you can find `tco` specific options at the top of this file https://github.com/llvm/llvm-project/blob/main/flang/tools/tco/tco.cpp#L41

For example, `./bin/bbc -emit-hlfir -o - file.f90 | ./bin/tco` is a bit like running `flang-new -fc1 -O2 -emit-llvm -o - file.f90`

File: flang/test/Lower/OpenMP/FIR/delayed_privatization.f90 (hunk 9 above)
Comment (ergawy) [reply flang/test/Lower/OpenMP/FIR/delayed_privatization.f90]: Added in https://github.com/llvm/llvm-project/pull/81452.

File: flang/test/Lower/OpenMP/FIR/delayed_privatization.f90 (hunk 11 above)
Comment (ergawy) [reply flang/test/Lower/OpenMP/FIR/delayed_privatization.f90]: Done in https://github.com/llvm/llvm-project/pull/81833 (I hope :))

**Comment by ergawy** [comment]
Abandoning since I am working on productizing the whole thing in separate PRs (see description of this PR for a detailed plan and tracking of that plan). I will keep updating the PR description to keep notes of the progress.
